### **Feature Engineering for Categorical Variables**

Models do arithmetic. Categories are labels. Every technique in this notebook is a way of bridging that gap, and each one makes a different claim about the data that can be either right or wrong.

That is the part worth holding onto. Encoding is not a formatting step you apply so the code stops erroring. Turning `Contract` into the numbers 0, 1 and 2 asserts that two year sits above one year by the same distance that one year sits above month-to-month. For `Contract` that assertion is defensible. Make it about `PaymentMethod` and you have told the model something false, and it will believe you.

**What we cover**

1. Reading a categorical column: ordinal against nominal, low cardinality against high
2. Label and ordinal encoding, measured, including what a wrong ordering costs
3. One-hot encoding, and where it stops working
4. Target encoding, the strongest technique here and the easiest to leak with
5. Frequency encoding and the hashing trick
6. Unseen categories and rare categories, which is where models break in production

### The dataset, and one engineered column

Telco is a good fit for this notebook because almost every column in it is categorical. What it lacks is a high-cardinality column, and the interesting encoding problems only appear once a column has more categories than you would want to one-hot.

So we build one. Each customer subscribes to some combination of nine services, and that combination describes their account in a way no single column does. Joining those nine columns into one string gives us a genuine high-cardinality feature, and it is the kind of thing you would actually construct at work rather than a synthetic column invented for a demo.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from sklearn.compose import ColumnTransformer
from sklearn.ensemble import RandomForestClassifier
from sklearn.feature_extraction import FeatureHasher
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import StratifiedKFold, cross_val_score, train_test_split
from sklearn.metrics import roc_auc_score
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import OneHotEncoder, OrdinalEncoder, StandardScaler, TargetEncoder

plt.style.use('seaborn-v0_8-whitegrid')
pd.set_option('display.max_columns', 30)
pd.set_option('display.width', 170)

RANDOM_STATE = 42

In [2]:
telco = pd.read_csv('datasets/telco_churn.csv')
telco['TotalCharges'] = pd.to_numeric(telco['TotalCharges'], errors='coerce')
telco = telco[telco['tenure'] > 0].reset_index(drop=True)

churned = (telco['Churn'] == 'Yes').astype(int)
NUMERIC = ['tenure', 'MonthlyCharges', 'TotalCharges']

# Build one high-cardinality column out of the nine service columns
SERVICES = ['PhoneService', 'MultipleLines', 'InternetService', 'OnlineSecurity',
            'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies']
telco['bundle'] = telco[SERVICES].agg('|'.join, axis=1)

print(f'Rows          : {len(telco):,}')
print(f'Churn rate    : {churned.mean():.1%}')
print(f'Unique bundles: {telco["bundle"].nunique()}')
print()
print('Two example bundles:')
for b in telco['bundle'].value_counts().head(2).index:
    print(f'  {b}')

Rows          : 7,032
Churn rate    : 26.6%
Unique bundles: 322

Two example bundles:
  Yes|No|No|No internet service|No internet service|No internet service|No internet service|No internet service|No internet service
  Yes|Yes|No|No internet service|No internet service|No internet service|No internet service|No internet service|No internet service


### 1. Reading a categorical column before encoding it

Two questions decide the method, and answering them takes longer than the encoding itself.

**Does the order mean anything?** An **ordinal** column has a genuine ranking. `Contract` runs month-to-month, one year, two year, and that sequence describes increasing commitment. A **nominal** column does not. `PaymentMethod` has four values and no defensible ordering among them.

Pandas cannot answer this for you. Both arrive as strings and look identical to the computer. The distinction lives in your understanding of the domain, which is why notebook 1 declared `Contract` as an ordered `Categorical` the moment we knew it.

**How many distinct values are there?** Low cardinality means one-hot encoding is comfortable. High cardinality means it is not, and the alternatives all involve trade-offs.

In [3]:
categorical_cols = [c for c in telco.columns
                    if c not in NUMERIC + ['customerID', 'Churn', 'bundle']]

inventory = pd.DataFrame({
    'column': categorical_cols,
    'n_categories': [telco[c].nunique() for c in categorical_cols],
    'most_common': [telco[c].value_counts().index[0] for c in categorical_cols],
    'rarest_count': [telco[c].value_counts().iloc[-1] for c in categorical_cols],
})
inventory.loc[len(inventory)] = ['bundle (engineered)', telco['bundle'].nunique(),
                                 telco['bundle'].value_counts().index[0][:28] + '...',
                                 telco['bundle'].value_counts().iloc[-1]]

print(inventory.to_string(index=False))

             column  n_categories                     most_common  rarest_count
             gender             2                            Male          3483
      SeniorCitizen             2                               0          1142
            Partner             2                              No          3393
         Dependents             2                              No          2099
       PhoneService             2                             Yes           680
      MultipleLines             3                              No           680
    InternetService             3                     Fiber optic          1520
     OnlineSecurity             3                              No          1520
       OnlineBackup             3                              No          1520
   DeviceProtection             3                              No          1520
        TechSupport             3                              No          1520
        StreamingTV             3       

Fifteen original columns hold between 2 and 4 categories each. The engineered `bundle` holds 322.

That gap is the whole subject of this notebook. Every original column can be one-hot encoded without a second thought, producing at most 4 new columns. Doing the same to `bundle` produces 322, and we will see shortly that this makes the model worse rather than better.

Before encoding anything, it is worth checking whether a column carries signal at all, because a column that does not predict the target does not deserve careful treatment.

In [4]:
bundle_stats = (pd.DataFrame({'bundle': telco['bundle'], 'churned': churned})
                .groupby('bundle')['churned'].agg(['mean', 'size'])
                .rename(columns={'mean': 'churn_rate', 'size': 'customers'}))

common = bundle_stats[bundle_stats['customers'] >= 20].sort_values('churn_rate')

print(f'Bundles with at least 20 customers: {len(common)}')
print(f'Their churn rates run from {common["churn_rate"].min():.1%} to {common["churn_rate"].max():.1%}')
print(f'(the dataset average is {churned.mean():.1%})')
print()
print('Lowest risk bundles:')
print(common.head(3).round(3).to_string())
print()
print('Highest risk bundles:')
print(common.tail(3).round(3).to_string())

Bundles with at least 20 customers: 76
Their churn rates run from 0.0% to 73.9%
(the dataset average is 26.6%)

Lowest risk bundles:
                                                 churn_rate  customers
bundle                                                                
Yes|Yes|DSL|Yes|Yes|Yes|Yes|No|No                       0.0         21
Yes|Yes|DSL|No|Yes|Yes|Yes|Yes|Yes                      0.0         25
No|No phone service|DSL|Yes|Yes|Yes|Yes|Yes|Yes         0.0         25

Highest risk bundles:
                                         churn_rate  customers
bundle                                                        
Yes|Yes|Fiber optic|No|No|No|No|Yes|Yes       0.664        110
Yes|Yes|Fiber optic|No|No|No|No|No|Yes        0.727         77
Yes|No|Fiber optic|No|No|No|No|Yes|No         0.739         69


Churn ranges from 0% to 73.9% depending on which services a customer has, against a 26.6% average. This column is worth encoding well.

It also shows the reason a frequency count matters. Some bundles appear over a thousand times and some appear twice, and a churn rate measured from two customers is not the same kind of evidence as one measured from a thousand. Section 4 comes back to this.

In [5]:
cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=RANDOM_STATE)

def logreg_auc(X):
    # Scaled logistic regression: sensitive to how categories are encoded
    return cross_val_score(make_pipeline(StandardScaler(), LogisticRegression(max_iter=3000)),
                           X, churned, cv=cv, scoring='roc_auc').mean()

def forest_auc(X):
    return cross_val_score(RandomForestClassifier(n_estimators=150, random_state=RANDOM_STATE, n_jobs=1),
                           X, churned, cv=cv, scoring='roc_auc').mean()


numeric_only = telco[NUMERIC]
BASELINE = logreg_auc(numeric_only)
print(f'Baseline, the three numeric columns and nothing else: AUC {BASELINE:.4f}')
print('Every encoding below gets compared against this number.')

Baseline, the three numeric columns and nothing else: AUC 0.8087
Every encoding below gets compared against this number.


### 2. Label and ordinal encoding

The simplest approach assigns each category an integer. `Bank transfer` becomes 0, `Credit card` becomes 1, and so on.

When the column is genuinely ordered this is exactly right, and it is called **ordinal encoding**. When the column is not ordered it is usually called **label encoding**, and it quietly inserts a ranking that does not exist.

Note what scikit-learn does when you do not specify an order: it sorts alphabetically. So the ranking your model receives is decided by spelling.

In [6]:
payment_categories = sorted(telco['PaymentMethod'].unique())

print('The order sklearn assigns by default, purely alphabetical:')
for i, cat in enumerate(payment_categories):
    print(f'  {i} = {cat}')
print()
print('That numbering claims Mailed check (3) is three units "more" than Bank transfer (0),')
print('and that Credit card (1) sits between them. None of that means anything.')

The order sklearn assigns by default, purely alphabetical:
  0 = Bank transfer (automatic)
  1 = Credit card (automatic)
  2 = Electronic check
  3 = Mailed check

That numbering claims Mailed check (3) is three units "more" than Bank transfer (0),
and that Credit card (1) sits between them. None of that means anything.


In [7]:
label_encoded = numeric_only.assign(
    payment=OrdinalEncoder().fit_transform(telco[['PaymentMethod']]).ravel())

one_hot = pd.concat([numeric_only,
                     pd.get_dummies(telco['PaymentMethod'], prefix='pay').astype(int)], axis=1)

print(f'{"encoding":16s} {"columns":>8s} {"LogReg AUC":>12s} {"RandomForest AUC":>18s}')
print('-' * 58)
print(f'{"label (0-3)":16s} {label_encoded.shape[1]:>8d} {logreg_auc(label_encoded):>12.4f} {forest_auc(label_encoded):>18.4f}')
print(f'{"one-hot":16s} {one_hot.shape[1]:>8d} {logreg_auc(one_hot):>12.4f} {forest_auc(one_hot):>18.4f}')

encoding          columns   LogReg AUC   RandomForest AUC
----------------------------------------------------------


label (0-3)             4       0.8088             0.7827


one-hot                 7       0.8175             0.7795


Label encoding costs the logistic regression 0.0087 AUC, dropping from 0.8175 to 0.8088. The random forest is barely affected, and in fact does marginally better with the single label column.

That split is the point, and it explains a disagreement you will see in practice.

A **linear model** multiplies the encoded column by one coefficient. With payment method encoded 0 to 3, it can only learn a single effect that grows or shrinks steadily as the number rises. Since the ordering is arbitrary, no such steady effect exists, and the model is forced to fit a straight line through noise.

A **tree** splits repeatedly, so it can isolate the value 2 with two cuts and largely recover the original categories. Trees are much more tolerant of label encoding, which is why gradient boosting practitioners often use it without much harm.

Now the extreme version of the same mistake.

In [8]:
label_bundle = numeric_only.assign(
    bundle=OrdinalEncoder().fit_transform(telco[['bundle']]).ravel())

print(f'Bundle label-encoded as integers 0 to {telco["bundle"].nunique() - 1}:')
print(f'  LogReg AUC       {logreg_auc(label_bundle):.4f}')
print(f'  Baseline (no bundle at all) {BASELINE:.4f}')
print()
print(f'  RandomForest AUC {forest_auc(label_bundle):.4f}')

Bundle label-encoded as integers 0 to 321:
  LogReg AUC       0.8087
  Baseline (no bundle at all) 0.8087



  RandomForest AUC 0.7940


The logistic regression scores 0.8087, identical to the baseline that does not use the bundle at all.

A column carrying churn rates from 0% to 73.9% contributed exactly nothing, because we handed it to the model as an arbitrary integer between 0 and 321 sorted alphabetically. The information was destroyed by the encoding, not absent from the data.

The random forest reaches 0.7940 and does extract something, since it can carve the integer line into pieces. It is still a poor use of the column.

Now `Contract`, where ordering is real. Notebook 1 measured churn at 42.7% for month-to-month, 11.3% for one year, and 2.8% for two year, which is a clean monotone decline. Ordinal encoding should be able to capture that in a single column.

In [9]:
correct_order = ['Month-to-month', 'One year', 'Two year']
wrong_order = ['One year', 'Two year', 'Month-to-month']    # alphabetical-ish, and meaningless

def encode_with(order):
    return numeric_only.assign(contract=telco['Contract'].map({c: i for i, c in enumerate(order)}))


ordinal_right = encode_with(correct_order)
ordinal_wrong = encode_with(wrong_order)
contract_onehot = pd.concat(
    [numeric_only, pd.get_dummies(telco['Contract'], prefix='contract').astype(int)], axis=1)

print(f'{"encoding":34s} {"columns":>8s} {"LogReg AUC":>12s}')
print('-' * 56)
print(f'{"ordinal, correct order":34s} {ordinal_right.shape[1]:>8d} {logreg_auc(ordinal_right):>12.4f}')
print(f'{"ordinal, meaningless order":34s} {ordinal_wrong.shape[1]:>8d} {logreg_auc(ordinal_wrong):>12.4f}')
print(f'{"one-hot":34s} {contract_onehot.shape[1]:>8d} {logreg_auc(contract_onehot):>12.4f}')

encoding                            columns   LogReg AUC
--------------------------------------------------------
ordinal, correct order                    4       0.8278
ordinal, meaningless order                4       0.8197
one-hot                                   6       0.8276


Three results worth separating.

**Correct ordinal encoding reaches 0.8278 using one column.** One-hot reaches 0.8276 using three. They tie, and the ordinal version does it with a third of the columns, because the ordering we supplied was real information the model did not have to learn.

**The wrong ordering costs 0.0081 AUC.** Same technique, same number of columns, and the only difference is that we shuffled which category got which integer. Ordinal encoding is only as good as the order you give it, and getting it wrong is worse than not claiming an order at all.

**One-hot never had to be told.** It makes no ordering claim, so it cannot make a wrong one, and it recovered the pattern from data alone. That safety is what you pay two extra columns for.

The practical guidance:

- Use **ordinal encoding** when you know the order, and write the order out explicitly rather than relying on the default sort.
- Use **one-hot** when there is no order, or when you are unsure.
- Never let alphabetical sorting decide the ranking, which is what `OrdinalEncoder` and `LabelEncoder` do if you say nothing.
- `LabelEncoder` is documented for encoding the **target**, not features. Use `OrdinalEncoder` for feature columns, since it accepts 2D input and fits into pipelines.

### 3. One-hot encoding, and where it stops working

One-hot creates one binary column per category. A customer on a two year contract gets a 1 in `contract_Two year` and 0 elsewhere. No ordering is implied, and each category gets its own coefficient.

It is the default for nominal columns for good reason. It is also the technique people keep using after it has stopped being appropriate.

In [10]:
sample = telco[['PaymentMethod']].head(4)

encoder = OneHotEncoder(sparse_output=False, dtype=int)
encoded = encoder.fit_transform(sample)

print('Input, the first four customers:')
print(sample.to_string(index=False))
print()
print('Output:')
print(pd.DataFrame(encoded, columns=encoder.get_feature_names_out()).to_string(index=False))
print()
print(f'Columns created : {encoded.shape[1]}')
print(f'Categories in the full dataset: {telco["PaymentMethod"].nunique()}')

Input, the first four customers:
            PaymentMethod
         Electronic check
             Mailed check
             Mailed check
Bank transfer (automatic)

Output:
 PaymentMethod_Bank transfer (automatic)  PaymentMethod_Electronic check  PaymentMethod_Mailed check
                                       0                               1                           0
                                       0                               0                           1
                                       0                               0                           1
                                       1                               0                           0

Columns created : 3
Categories in the full dataset: 4


Look at what just happened, because it is the whole argument for using `OneHotEncoder` over `get_dummies`.

`PaymentMethod` has four categories. We fitted on four rows that happened to contain only three of them, and got three columns. `Credit card (automatic)` has no column at all, because the encoder never saw it.

That is fine here, since we only fitted on a sample. It becomes a production outage when the same thing happens between training and serving.

`pd.get_dummies` inspects whatever data you hand it and creates columns for the categories present. Train on data with four payment methods and you get four columns; run it next month on a batch that happens to contain three and you get three columns, in a different order, and the model silently receives the wrong values in the wrong positions.

`OneHotEncoder` learns the category list once during `fit` and applies that same list forever, including categories absent from the data it is transforming. It belongs in a pipeline for exactly the same reason every other transformer does. Use `get_dummies` for quick exploration and `OneHotEncoder` for anything that will run twice.

Now the failure case that no amount of correct plumbing fixes.

In [11]:
bundle_onehot = pd.concat(
    [numeric_only, pd.get_dummies(telco['bundle'], prefix='b').astype(int)], axis=1)

print(f'{"approach":34s} {"columns":>8s} {"LogReg AUC":>12s}')
print('-' * 56)
print(f'{"numeric only, bundle ignored":34s} {numeric_only.shape[1]:>8d} {BASELINE:>12.4f}')
print(f'{"bundle one-hot encoded":34s} {bundle_onehot.shape[1]:>8d} {logreg_auc(bundle_onehot):>12.4f}')

approach                            columns   LogReg AUC
--------------------------------------------------------
numeric only, bundle ignored              3       0.8087


bundle one-hot encoded                  325       0.8007


One-hot encoding the bundle made the model **worse** than ignoring the column entirely, dropping from 0.8087 to 0.8007.

We already know the column is informative, so this is a failure of the encoding rather than the data. Three things are going wrong at once:

- **Each column is almost entirely zeros.** With 322 categories, the average column is 1 for about 22 rows out of 7,032. There is very little for a coefficient to be estimated from.
- **Rare categories produce unreliable coefficients.** A bundle appearing twice gets its own column and its own parameter, fitted from two observations. That parameter is noise, and the model will use it anyway.
- **325 columns against 7,032 rows** gives the model far more freedom than the data can support, so it fits patterns in the training folds that do not survive to the validation fold.

This is the point where you need something other than one-hot. Sections 4 and 5 cover the alternatives, and section 6 shows that one-hot can be rescued if you deal with the rare categories first.

### 4. Target encoding, and the leak that comes with it

Target encoding replaces each category with the average target value for that category. Every bundle becomes the churn rate of the customers who have it, so 322 categories collapse into one numeric column that is directly meaningful.

It is the strongest technique in this notebook for high-cardinality columns. It is also the one that will silently give you a beautiful validation score and a model that fails in production, so we are going to do it wrong first and measure the damage.

In [12]:
# THE WRONG WAY: compute the mapping from the whole dataset, then cross-validate
leaky_map = churned.groupby(telco['bundle']).mean()
leaky = numeric_only.assign(bundle_te=telco['bundle'].map(leaky_map))

print('Each bundle replaced by its churn rate, computed over all 7,032 rows:')
print(leaky[['bundle_te']].head(3).round(4).to_string(index=False))
print()
print(f'Cross-validated AUC: {logreg_auc(leaky):.4f}')
print(f'Baseline           : {BASELINE:.4f}')
print()
print('That looks like a large improvement. It is not real.')

Each bundle replaced by its churn rate, computed over all 7,032 rows:
 bundle_te
    0.3871
    0.1111
    0.2051

Cross-validated AUC: 0.8450
Baseline           : 0.8087

That looks like a large improvement. It is not real.


The problem is subtle enough that it survives code review regularly.

`cross_val_score` splits the data into five folds and holds one out each time. But the mapping was built **before** the split, using every row including the ones about to be held out. When the model scores fold 3, the encoded value it reads for each customer was computed partly from that customer's own churn outcome.

The effect is worst for rare categories. A bundle that appears twice takes the average of those two customers' outcomes, so if both churned the encoded value is 1.0, and the model reads that value for those exact customers at validation time. It is not learning a pattern, it is reading the answer.

Two fixes exist and they solve different halves of the problem. The first is the familiar one: fit inside the pipeline so the mapping only ever sees training rows. The second is specific to target encoding: even within the training set, a row should not contribute to the mapping used to encode itself, which requires an internal cross-fitting scheme. `sklearn.preprocessing.TargetEncoder` does both.

In [13]:
# THE RIGHT WAY: TargetEncoder inside the pipeline, cross-fitted internally
preprocess = ColumnTransformer([
    ('numeric', 'passthrough', NUMERIC),
    ('target_encode', TargetEncoder(target_type='binary', random_state=RANDOM_STATE), ['bundle']),
])

proper = make_pipeline(preprocess, StandardScaler(), LogisticRegression(max_iter=3000))
proper_auc = cross_val_score(proper, telco[NUMERIC + ['bundle']], churned,
                             cv=cv, scoring='roc_auc').mean()

print(f'{"approach":38s} {"columns":>8s} {"AUC":>9s}')
print('-' * 58)
print(f'{"numeric only":38s} {3:>8d} {BASELINE:>9.4f}')
print(f'{"bundle one-hot":38s} {325:>8d} {logreg_auc(bundle_onehot):>9.4f}')
print(f'{"bundle target-encoded, LEAKED":38s} {4:>8d} {logreg_auc(leaky):>9.4f}')
print(f'{"bundle target-encoded, PROPER":38s} {4:>8d} {proper_auc:>9.4f}')

approach                                columns       AUC
----------------------------------------------------------
numeric only                                  3    0.8087
bundle one-hot                              325    0.8007


bundle target-encoded, LEAKED                 4    0.8450
bundle target-encoded, PROPER                 4    0.8250


The leak inflated the score by 0.0200 AUC, from a true 0.8250 to a reported 0.8450.

Sit with what that means in practice. You would have reported 0.845 to your team, planned around it, and then watched the deployed model behave like a 0.825 model. The gap is not large enough to look suspicious, which is exactly what makes it dangerous. A leak that produced 0.99 would get caught immediately.

The honest result is still excellent. Proper target encoding reaches 0.8250 using **one column**, against one-hot's 0.8007 using 322, and against 0.8087 for ignoring the bundle. That is the largest single-column gain anywhere in this series so far.

To make the leak concrete rather than statistical, here it is on a genuine held-out test set.

In [14]:
train, test, y_train, y_test = train_test_split(
    telco, churned, test_size=0.25, random_state=RANDOM_STATE, stratify=churned)

# Version A: mapping built from ALL rows, including the test set
map_all = churned.groupby(telco['bundle']).mean()
model_a = make_pipeline(StandardScaler(), LogisticRegression(max_iter=3000)).fit(
    train[NUMERIC].assign(te=train['bundle'].map(map_all)), y_train)
auc_a = roc_auc_score(y_test, model_a.predict_proba(
    test[NUMERIC].assign(te=test['bundle'].map(map_all)))[:, 1])

# Version B: mapping built from TRAINING rows only, unseen bundles fall back to the prior
map_train = y_train.groupby(train['bundle']).mean()
prior = y_train.mean()
model_b = make_pipeline(StandardScaler(), LogisticRegression(max_iter=3000)).fit(
    train[NUMERIC].assign(te=train['bundle'].map(map_train).fillna(prior)), y_train)
auc_b = roc_auc_score(y_test, model_b.predict_proba(
    test[NUMERIC].assign(te=test['bundle'].map(map_train).fillna(prior)))[:, 1])

print(f'Mapping built from all rows,      test AUC = {auc_a:.4f}   <- flattering and false')
print(f'Mapping built from training only, test AUC = {auc_b:.4f}   <- what you would actually get')

Mapping built from all rows,      test AUC = 0.8418   <- flattering and false
Mapping built from training only, test AUC = 0.8153   <- what you would actually get


0.8418 against 0.8153. The first number is not a better model, it is the same model being graded on an exam it had already seen.

#### Smoothing, for categories with too little evidence

One more issue remains even with the leakage handled. A bundle seen 3 times where nobody churned has an observed rate of 0.0, and taking that at face value tells the model those customers never leave. Three observations do not support that claim.

Smoothing blends each category's own average toward the overall average, weighted by how much evidence the category has:

$$\text{encoded} = \frac{n \times \text{category mean} + m \times \text{overall mean}}{n + m}$$

`n` is how many rows the category has and `m` controls how much pull the overall mean gets. Categories with plenty of rows barely move. Categories with almost no rows get pulled most of the way to the prior.

In [15]:
overall = churned.mean()
counts = telco['bundle'].value_counts()

examples = [counts.index[0], counts.index[1], counts[counts <= 3].index[0]]

print(f'Overall churn rate (the prior) = {overall:.3f}')
print()
print(f'{"customers":>10s} {"raw mean":>10s} {"m=0":>8s} {"m=10":>8s} {"m=50":>8s}')
print('-' * 48)
for bundle in examples:
    rows = telco['bundle'] == bundle
    n = rows.sum()
    raw = churned[rows].mean()
    smoothed = [(n * raw + m * overall) / (n + m) for m in (0, 10, 50)]
    print(f'{n:>10d} {raw:>10.3f} ' + ' '.join(f'{s:>8.3f}' for s in smoothed))

Overall churn rate (the prior) = 0.266

 customers   raw mean      m=0     m=10     m=50
------------------------------------------------
      1180      0.087    0.087    0.089    0.095
       340      0.029    0.029    0.036    0.060
         3      0.000    0.000    0.204    0.251


The bundle with 1,180 customers barely moves, from 0.087 to 0.095 even at `m=50`. It has enough evidence to speak for itself.

The bundle with 3 customers moves from 0.000 to 0.251, almost all the way to the 0.266 prior. With three observations, the honest encoding is close to "we do not know anything specific about this group".

`TargetEncoder` applies smoothing automatically with `smooth='auto'`, which is why we did not have to think about it above. Knowing the mechanism still matters, because when you see a target-encoded column whose values all sit suspiciously close to the overall mean, smoothing is usually the reason.

When to reach for target encoding:

- High-cardinality nominal columns, where one-hot produces too many columns.
- When you have enough rows per category for the averages to mean something.
- Always inside a pipeline, always cross-fitted, and always with smoothing.

When to avoid it:

- Small datasets, where every category is rare and the encoding is mostly noise.
- Any column close to an identifier, where target encoding becomes a direct route to memorising the target. Notebook 1's `customerID` would be the extreme case.

### 5. Frequency encoding and the hashing trick

Two more options for high-cardinality columns, both of which ignore the target entirely and therefore cannot leak.

**Frequency encoding** replaces each category with how often it appears. It is a single column, needs no target, and the implicit claim is that how common something is might itself be predictive.

In [16]:
frequencies = telco['bundle'].map(telco['bundle'].value_counts() / len(telco))
freq_encoded = numeric_only.assign(bundle_freq=frequencies)

print(f'Frequency encoded : AUC {logreg_auc(freq_encoded):.4f}')
print(f'Baseline          : AUC {BASELINE:.4f}')
print(f'Target encoded    : AUC {proper_auc:.4f}')

Frequency encoded : AUC 0.8094
Baseline          : AUC 0.8087
Target encoded    : AUC 0.8250


0.8094 against a 0.8087 baseline. Essentially nothing, and that is worth reporting rather than skipping.

Frequency encoding only helps when rarity itself carries signal, which is true for things like fraud detection, where unusual merchant categories are genuinely suspicious. Here, how common a service bundle is has almost no relationship to whether the customer leaves, so the encoding faithfully captures a variable that does not matter.

The technique is cheap, safe, and worth trying. It is not a substitute for target encoding.

**The hashing trick** takes a different route. It runs each category through a hash function and uses the result to pick a column, mapping any number of categories into a fixed number of columns you choose in advance. Different categories can collide into the same column, and the bet is that collisions are rare enough and random enough not to matter.

In [17]:
print(f'{"columns":>9s} {"categories per bucket":>24s} {"LogReg AUC":>12s}')
print('-' * 48)
for n_features in [8, 16, 32, 64]:
    hasher = FeatureHasher(n_features=n_features, input_type='string')
    hashed = pd.DataFrame(hasher.transform([[b] for b in telco['bundle']]).toarray(),
                          index=telco.index).add_prefix('hash_')
    score = logreg_auc(pd.concat([numeric_only, hashed], axis=1))
    print(f'{n_features:>9d} {telco["bundle"].nunique() / n_features:>24.1f} {score:>12.4f}')

print('-' * 48)
print(f'{325:>9d} {1.0:>24.1f} {logreg_auc(bundle_onehot):>12.4f}   <- one-hot, for comparison')

  columns    categories per bucket   LogReg AUC
------------------------------------------------


        8                     40.2       0.8087
       16                     20.1       0.8107
       32                     10.1       0.8142
       64                      5.0       0.8156
------------------------------------------------


      325                      1.0       0.8007   <- one-hot, for comparison


Hashing 322 categories into 64 columns scores 0.8156, which beats one-hot's 0.8007 using a fifth of the columns.

That is initially surprising, since hashing deliberately throws information away by letting categories collide. The explanation is that one-hot was not suffering from too little information, it was suffering from too many parameters estimated from too few rows each. Hashing forces categories to share columns, each column ends up with far more non-zero rows, and the coefficients become estimable. The collisions cost less than the sparsity did.

Where hashing genuinely earns its place is scale. With 500,000 user IDs, one-hot is impossible and target encoding needs a 500,000-entry lookup table shipped alongside the model. A hasher needs no stored vocabulary at all, since the hash function is fixed, so it handles categories it has never seen without any special case and uses constant memory.

The cost is interpretability. Column `hash_37` holds a mixture of unrelated categories and means nothing to a human, so feature importance on hashed columns cannot be explained. Use it when cardinality is genuinely enormous, and prefer target encoding when it is merely large.

### 6. Unseen and rare categories

The two situations that break models after deployment, both of which are invisible while you are working with a single fixed dataset.

**Unseen categories** are values that appear at prediction time and were not in the training data. A new payment method launches, a new product category is added, a new city gets a warehouse. Your encoder has never seen it and has to do something.

In [18]:
# How many unseen bundles appear depends entirely on which split you happen to draw
print(f'{"test size":>10s} {"seed":>6s} {"unseen bundles":>16s} {"rows affected":>15s}')
print('-' * 52)
for test_size in [0.25, 0.40]:
    for seed in range(4):
        tr, te = train_test_split(telco, test_size=test_size, random_state=seed, stratify=churned)
        unseen = set(te['bundle']) - set(tr['bundle'])
        print(f'{test_size:>10.2f} {seed:>6d} {len(unseen):>16d} {te["bundle"].isin(unseen).sum():>15d}')

 test size   seed   unseen bundles   rows affected
----------------------------------------------------
      0.25      0                2               4


      0.25      1                0               0
      0.25      2                2               5
      0.25      3                2               5
      0.40      0                8              24
      0.40      1                3               6
      0.40      2                8              22
      0.40      3                7              16


Between 0 and 8 unseen categories depending on nothing but the random seed.

That variability is the actual lesson. Draw seed 1 at a 25% test size and you see no unseen categories at all, conclude your encoding is fine, and ship it. Draw seed 0 and the problem is visible. **Whether you discover this bug during development is a matter of luck**, which is why it has to be handled by design rather than by testing until it shows up.

Since we cannot rely on a lucky split to demonstrate the behaviour, here it is with a category that is guaranteed not to exist.

In [19]:
encoder_input = telco[['bundle']]
brand_new = pd.DataFrame({'bundle': ['Yes|Yes|Satellite|Yes|Yes|Yes|Yes|Yes|Yes']})   # a service we never sold

print('Fitting on all 322 known bundles, then asking about a bundle that has never existed.')
print()

try:
    OneHotEncoder(handle_unknown='error').fit(encoder_input).transform(brand_new)
except ValueError as err:
    print(f"handle_unknown='error'   -> ValueError: {str(err)[:70]}...")

ignored = OneHotEncoder(handle_unknown='ignore', sparse_output=False).fit(encoder_input).transform(brand_new)
print(f"handle_unknown='ignore'  -> all {ignored.shape[1]} columns are zero: {bool((ignored == 0).all())}")

infreq = OneHotEncoder(min_frequency=20, handle_unknown='infrequent_if_exist',
                       sparse_output=False).fit(encoder_input)
routed = infreq.transform(brand_new)
print(f"handle_unknown='infrequent_if_exist' -> routed into the rare bucket, "
      f"column {int(np.argmax(routed))} of {routed.shape[1]}")

te_encoder = TargetEncoder(target_type='binary', random_state=RANDOM_STATE).fit(encoder_input, churned)
print(f"TargetEncoder            -> unseen category gets the overall mean: "
      f"{te_encoder.transform(brand_new)[0][0]:.4f}  (dataset rate {churned.mean():.4f})")

Fitting on all 322 known bundles, then asking about a bundle that has never existed.

handle_unknown='error'   -> ValueError: Found unknown categories ['Yes|Yes|Satellite|Yes|Yes|Yes|Yes|Yes|Yes']...
handle_unknown='ignore'  -> all 322 columns are zero: True
handle_unknown='infrequent_if_exist' -> routed into the rare bucket, column 76 of 77
TargetEncoder            -> unseen category gets the overall mean: 0.2658  (dataset rate 0.2658)


Four different answers to the same question, and none is automatically right:

- **`handle_unknown='error'`** fails loudly. Appropriate when an unknown category means something has genuinely gone wrong upstream and you want to know before serving predictions.
- **`handle_unknown='ignore'`** encodes the row as all zeros. The model still returns a prediction, based only on the other features. This is usually what you want for a service that must stay up, and it is worth understanding that "all zeros" is a category the model never saw during training either.
- **`handle_unknown='infrequent_if_exist'`** routes unknown values into the same bucket as rare training categories, which is more principled than all-zeros because that bucket has actually been trained on.
- **`TargetEncoder`** returns the overall mean, 0.2658, which is exactly the dataset churn rate. Knowing nothing specific about a category, predict the average, which is the right default.

**Rare categories** are the related problem. They exist in training, but with so few rows that anything learned about them is noise. Section 3 showed one-hot encoding the bundle scoring worse than ignoring it, and rare categories were a large part of why.

In [20]:
counts = telco['bundle'].value_counts()

print(f'{"min rows to keep":>17s} {"categories":>12s} {"columns":>9s} {"LogReg AUC":>12s}')
print('-' * 54)
for threshold in [1, 5, 20, 50, 100]:
    keep = counts[counts >= threshold].index
    grouped = telco['bundle'].where(telco['bundle'].isin(keep), 'Other')
    X = pd.concat([numeric_only, pd.get_dummies(grouped, prefix='b').astype(int)], axis=1)
    print(f'{threshold:>17d} {grouped.nunique():>12d} {X.shape[1]:>9d} {logreg_auc(X):>12.4f}')

 min rows to keep   categories   columns   LogReg AUC
------------------------------------------------------
                1          322       325       0.8007


                5          266       269       0.8045
               20           77        80       0.8238
               50           23        26       0.8235
              100           10        13       0.8168


This is the result to remember from this section.

One-hot encoding all 322 bundles scored 0.8007, worse than ignoring the column. Grouping everything below 20 rows into a single `Other` category leaves 77 categories, and the score jumps to 0.8238. We deleted distinctions between 245 categories and the model got substantially better.

The pattern has a clear peak. Grouping too little (threshold 5, 266 categories) barely helps at 0.8045. Grouping too much (threshold 100, 10 categories) falls back to 0.8168, because now genuinely different bundles are being merged. The threshold is a hyperparameter, and 20 to 50 happens to be the useful range here.

`OneHotEncoder` can do this directly, which keeps it inside the pipeline where it belongs.

In [21]:
for min_freq in [20, 50]:
    encoder = OneHotEncoder(min_frequency=min_freq, handle_unknown='infrequent_if_exist',
                            sparse_output=False)
    encoded = pd.DataFrame(encoder.fit_transform(telco[['bundle']]),
                           index=telco.index).add_prefix('b_')
    score = logreg_auc(pd.concat([numeric_only, encoded], axis=1))
    print(f'OneHotEncoder(min_frequency={min_freq:3d}) -> {encoded.shape[1]:3d} columns, AUC {score:.4f}')

print()
print('Same result as the manual grouping, and it survives into production,')
print('because unseen categories are routed to the infrequent bucket automatically.')

OneHotEncoder(min_frequency= 20) ->  77 columns, AUC 0.8238
OneHotEncoder(min_frequency= 50) ->  23 columns, AUC 0.8235

Same result as the manual grouping, and it survives into production,
because unseen categories are routed to the infrequent bucket automatically.


### 7. Putting the comparison together

Every encoding in this notebook, applied to the same column and scored the same way.

In [22]:
summary = pd.DataFrame([
    ('ignore the column',        3,   BASELINE),
    ('label encoding',           4,   logreg_auc(label_bundle)),
    ('one-hot, all 322',         325, logreg_auc(bundle_onehot)),
    ('frequency encoding',       4,   logreg_auc(freq_encoded)),
    ('hashing, 64 columns',      67,  logreg_auc(pd.concat([numeric_only, pd.DataFrame(
        FeatureHasher(n_features=64, input_type='string').transform(
            [[b] for b in telco['bundle']]).toarray(),
        index=telco.index).add_prefix('hash_')], axis=1))),
    ('one-hot, rare grouped',    80,  logreg_auc(pd.concat([numeric_only, pd.get_dummies(
        telco['bundle'].where(telco['bundle'].isin(counts[counts >= 20].index), 'Other'),
        prefix='b').astype(int)], axis=1))),
    ('target encoding, proper',  4,   proper_auc),
    ('target encoding, LEAKED',  4,   logreg_auc(leaky)),
], columns=['encoding', 'columns', 'AUC']).sort_values('AUC').reset_index(drop=True)

print(summary.round(4).to_string(index=False))

               encoding  columns    AUC
       one-hot, all 322      325 0.8007
      ignore the column        3 0.8087
         label encoding        4 0.8087
     frequency encoding        4 0.8094
    hashing, 64 columns       67 0.8156
  one-hot, rare grouped       80 0.8238
target encoding, proper        4 0.8250
target encoding, LEAKED        4 0.8450


Reading that table from the bottom up:

The leaked target encoding sits at the top and is the only entry that is not a real result. Removing it, proper target encoding wins at 0.8250 using four columns total, followed by rare-grouped one-hot at 0.8238 using eighty.

Those two are close enough that the choice between them is not really about the score. Target encoding is more compact and needs careful cross-fitting to be safe. Grouped one-hot is more transparent, cannot leak, and each column still means something you can point at. On a production system I would want a strong reason before choosing the technique that fails silently over the one that cannot.

At the bottom, label encoding and unrestricted one-hot both land at or below the baseline. Both took an informative column and made it useless, in opposite ways: one destroyed the distinctions, the other kept all of them and gave the model nothing solid to estimate from.

### <font color='green'>Key things to remember from this notebook</font>

Encoding is a claim about the data, not a formatting step. Ordinal encoding asserts an ordering, and the assertion can be false. `Contract` encoded in its true order scored 0.8278 from one column, matching one-hot's 0.8276 from three, while the same technique with a shuffled order lost 0.0081. If you know the order, state it explicitly, because `OrdinalEncoder` and `LabelEncoder` otherwise sort alphabetically and let spelling decide the ranking.

Label encoding hurts linear models and mostly spares trees. On `PaymentMethod` it cost logistic regression 0.0087 while the random forest was unaffected, because a linear model gets one coefficient for the whole column and a tree can split it repeatedly. Label encoding the 322-category bundle scored 0.8087, exactly the baseline, destroying a column whose churn rates ran from 0% to 73.9%.

One-hot stops working as cardinality grows. All 322 bundles one-hot encoded scored 0.8007, worse than dropping the column, because 325 columns over 7,032 rows leaves each coefficient estimated from about 22 non-zero rows. Grouping everything under 20 rows into `Other` cut it to 77 categories and lifted the score to 0.8238.

Target encoding is the strongest option here and the easiest to leak with. Building the mapping before the split reported 0.8450 against a true 0.8250, and on a held-out test set the leaked version claimed 0.8418 where the honest one delivered 0.8153. That gap is small enough to look plausible, which is what makes it dangerous. Use `TargetEncoder` inside a pipeline so the mapping is both fitted on training rows only and cross-fitted within them.

Smooth target encodings toward the prior. A bundle with 3 customers and no churn encodes as 0.000 raw and 0.251 smoothed, against a 0.266 overall rate. Three observations do not support a claim of zero.

Hashing beat one-hot on this data, 0.8156 from 64 columns against 0.8007 from 322, because forcing categories to share columns gave each coefficient more rows to learn from. Reach for it when cardinality is genuinely enormous and you cannot ship a lookup table, and accept that the columns become uninterpretable.

Decide what happens to unseen categories before you deploy, not after. `handle_unknown='error'` fails loudly, `'ignore'` encodes all zeros, and `'infrequent_if_exist'` routes into the rare bucket that was actually trained on. Only 4 test rows were affected here because this is a fixed historical extract, and in a live system that number grows every month.

Notebook 5 moves to dates and times, where a single timestamp column hides a dozen features and the ordering of rows starts to matter in ways it has not so far.